Support free plans (fixed conflicts).

PetrDlouhy opened the pull request:
This is #36 with resolved conflicts.

Alir3z4 commented:
@PetrDlouhy could you please update the changelog as well ?

Alir3z4 commented:
@PetrDlouhy could you please check this pull request so I can merge ?

PetrDlouhy commented:
@Alir3z4 I have rebased this and added the Changelog note

PetrDlouhy commented:
I just realised, there might be unexpected behaviour connected with this issue. After any paid plan, one would expect, that the free plan quotas will be applied, but that doesn't happen.

I don't know what was the expected behaviour before this function, but this seems like something, that would need to be cleared up before merging this.

@Alir3z4  What behaviour would do you consider adequate and compatible with current use?

PetrDlouhy commented on plans/models.py:
```diff
@@ -259,19 +267,17 @@ def extend_account(self, plan, pricing):
                 # This should not ever happen (as this case should be managed by plan change request)
                 # but just in case we consider a case when user has a different plan
-                if self.expire is None:
+                if not self.plan.is_free and self.expire is None:
```
This is clearly a bug. The `is_free` is function and should be called with `()`. We should test this.

PetrDlouhy commented:
@Alir3z4 I have rebased this to current master and fixed few misbehaviours. Now it should have following logic:
- If user switches to free account, the `expiration` is set to `None`.
- If user switches from non free account without `expiration`, to another non free account, the `expiration=None` stays the same (this is the original behaviour of `django-plans`).
- If user switches from free account to non free account the expiration is set from today.

- It also fixes the behaviour when the `expire_plans` action doesn't run. Now the validators considers the expired plan (even when `active=True`)to be the default plan.

Alir3z4 commented:
@PetrDlouhy Thanks for the great work and getting back to this, I understand how annoying it could be to get back to something that has been a done a long time ago.

I looked over the code and the implementation looks good to me.

Free plan is a new feature and I'd highly recommend to write documentation for it and how it can be used, otherwise it will be a hidden feature.

Could you please update the documentation to show how it's done and how it can be used ?

PetrDlouhy commented:
@Alir3z4 I have added few notes about free plans to the docs. I also realized, that the `expire_accounts` task is not described there at all, so I added page about it.

I also copied management command, which I use instead of the Celery task in my project.

PetrDlouhy commented:
@Alir3z4 Is there anything, I can do to get this merged?
